### PR TITLE
Call RateLimiter for compaction reads

### PR DESCRIPTION
**Pull request by ajkr**
Allow users to rate limit background work based on read bytes, written bytes, or sum of read and written bytes. Support these by changing the RateLimiter API, so no additional options were needed.

Test Plan:

- `db_bench compact` on 1GB database with 1MB/s read rate limit and 1ms refill interval; verified reads are chunked (at most 1048 bytes per chunk), and verified rate limit is upheld using iostat. Command: `TEST_TMPDIR=/data/compaction_bench/  ./db_bench -benchmarks=compact -statistics -use_existing_db -compression_type=none -rate_limiter_bytes_per_sec=1048576 -rate_limit_reads`.
- `db_stress` with lots of compactions, 1MB/s read rate limit, and 1ms refill interval. Command: `TEST_TMPDIR=/data/rocksdb-test/ ./db_stress -rate_limiter_bytes_per_sec=1048576 -rate_limit_reads -cache_size=0 -writepercent=100 -readpercent=0 -prefixpercent=0 -delpercent=0 -iterpercent=0 -write_buffer_size=1048576`
- also `db_stress` in direct I/O mode: `TEST_TMPDIR=/data/rocksdb-test/ ./db_stress -rate_limiter_bytes_per_sec=4200000 -rate_limit_bg_reads -cache_size=0 -writepercent=100 -readpercent=0 -prefixpercent=0 -delpercent=0 -iterpercent=0 -use_direct_io_for_flush_and_compaction -max_bytes_for_level_base=51200000 -threads=1 -target_file_size_base=12800000 -write_buffer_size=12800000 -clear_column_family_one_in=0 -reopen=0 -column_families=1 -block_size=16384 -ops_per_thread=100000000 -max_background_compactions=8`
- `db_bench -benchmarks=fillrandom` to load 1B keys before and after the change, both without applying any rate limit. Verified throughput is exactly the same (24.0 MB/s).
- unit tests

### CODE REVIEW

File: util/file_reader_writer.cc [hunk 1]
```diff
@@ -318,21 +344,21 @@ Status WritableFileWriter::RangeSync(uint64_t offset, uint64_t nbytes) {
   return writable_file_->RangeSync(offset, nbytes);
 }
 
-size_t WritableFileWriter::RequestToken(size_t bytes, bool align) {
-  Env::IOPriority io_priority;
-  if (rate_limiter_ && (io_priority = writable_file_->GetIOPriority()) <
-      Env::IO_TOTAL) {
-    bytes = std::min(
-      bytes, static_cast<size_t>(rate_limiter_->GetSingleBurstBytes()));
+size_t RequestToken(size_t bytes, size_t alignment, Env::IOPriority io_priority,
```
Comment (lightmark) [inline util/file_reader_writer.cc]: Is it better to put this method in RateLimiter class?

File: util/file_reader_writer.cc [hunk 2]
```diff
@@ -99,7 +99,33 @@ Status RandomAccessFileReader::Read(uint64_t offset, size_t n, Slice* result,
       *result = Slice(scratch, r);
 #endif  // !ROCKSDB_LITE
     } else {
-      s = file_->Read(offset, n, result, scratch);
+      size_t pos = 0;
```
Comment (lightmark) [inline util/file_reader_writer.cc]: Direct IO needs to be supported.

File: tools/db_stress.cc [hunk 3]
```diff
@@ -326,6 +326,11 @@ DEFINE_double(max_bytes_for_level_multiplier, 2,
 DEFINE_int32(range_deletion_width, 10,
              "The width of the range deletion intervals.");
 
+DEFINE_uint64(rate_limiter_bytes_per_sec, 0, "Set options.rate_limiter value.");
+
+DEFINE_bool(rate_limit_reads, false,
```
Comment (lightmark) [inline tools/db_stress.cc]: rate_limit_bg_reads?

File: tools/db_bench_tool.cc [hunk 4]
```diff
@@ -785,6 +785,9 @@ DEFINE_int32(rate_limit_delay_max_milliseconds, 1000,
 
 DEFINE_uint64(rate_limiter_bytes_per_sec, 0, "Set options.rate_limiter value.");
 
+DEFINE_bool(rate_limit_reads, false,
```
Comment (lightmark) [inline tools/db_bench_tool.cc]: rate_limit_bg_reads?

File: util/file_reader_writer.cc [hunk 5]
```diff
@@ -99,7 +99,33 @@ Status RandomAccessFileReader::Read(uint64_t offset, size_t n, Slice* result,
...
+        } else {
+          // make sure chunks are inserted contiguously into `res_scratch`.
+          assert(tmp_result.data() == res_scratch + pos);
+        }
+        pos += tmp_result.size();
+        if (!s.ok() || tmp_result.size() < allowed) {
+          break;
+        }
+      }
+      *result = Slice(res_scratch, pos);
```
Comment (lightmark) [inline util/file_reader_writer.cc]: Do we need to copy the data from res_scratch to scratch if the file is mmap'd file?

**Comment by siying** [comment]
We need to think about whether rate limiting in the same way for reads and writes is the right thing to do. RocksDB's writes are mostly sequential, so limiting writes in the byte rate works for both of flash and hard drives. However, reads (except compaction reads) are more likely to be small random. In the case of hard drives (the instance use case we are targetting with this PR), using byte rate may not able to limit the I/O traffic in the way we want, because of the drive seeking effects. For example, a 128KB random read costs far less than 2X of 64KB random read, while it is much much more expensive than 128KB compaction writes. So for hard drives, may be number of I/Os is a better one to limit than byte rate. For flash, it is even more complicated.

**Comment by ajkr** [comment]
To clarify, this PR only affects compaction reads, which should have the ideal sequential I/O pattern as you mentioned.

File: tools/db_stress.cc (hunk 3 above)
Comment (ajkr) [inline tools/db_stress.cc]: sure

File: util/file_reader_writer.cc (hunk 5 above)
Comment (ajkr) [inline util/file_reader_writer.cc]: actually we should avoid copy for mmap'd files. for file reader we mmap with PROT_READ permission only, so should be safe.

File: util/file_reader_writer.cc (hunk 1 above)
Comment (ajkr) [inline util/file_reader_writer.cc]: but TruncateToPageBoundary isn't in include/ :(

Comment (lightmark) [inline util/file_reader_writer.cc]: maybe needs refactoring.

**Comment by siying** [comment]
@ajkr OK. Sorry for misunderstanding it. I think I was still suffering from jet lags yesterday. It's OK then. It's good as long as @lightmark approves it then.

File: util/file_reader_writer.cc (hunk 5 above)
Comment (lightmark) [inline util/file_reader_writer.cc]: if s.ok() == false, result length should be 0.

**Pull request closed without merging**